src/nagini_translation/translators/call.py

```diff
@@ -935,15 +964,344 @@ def translate_Call(self, node: ast.Call, ctx: Context, impure=False) -> StmtsAnd
+        assert isinstance(node.func, ast.Attribute)
+        thread_stmt, thread = self.translate_expr(node.func.value, ctx)
+        stmts = thread_stmt
+
+        # Assert that thread may be joined.
+        joinable_pos = self.to_position(node, ctx, rules=rules.THREAD_JOIN_JOINABLE)
+        joinable_func = self.viper.FuncApp(JOINABLE_FUNC, [thread],
+                                           joinable_pos, info,
+                                           self.viper.Bool)
+
+        stmts.append(self.viper.Assert(joinable_func, joinable_pos, info))
+
+        # Check that thread wait level is above current wait level (prevents deadlocks).
+        wait_level_pos = self.to_position(node, ctx, rules=rules.THREAD_JOIN_WAITLEVEL)
+        thread_level = self.create_level_call(sil.RefExpr(thread))
+        residue_var = sil.PermVar(ctx.actual_function.obligation_info.residue_level)
+        obligation_assertion = self.create_level_below(thread_level, residue_var, ctx)
+        obligation_assertion = obligation_assertion.translate(self, ctx, wait_level_pos,
+                                                              info)
+        stmts.append(self.viper.Assert(obligation_assertion, wait_level_pos, info))
+
+        # Check how much permission is held to ThreadPost. This amount of the thread
+        # postcondition will be inhaled later.
+        post_pred_acc = self.viper.PredicateAccess([thread], THREAD_POST_PRED, pos, info)
+        post_perm = self.viper.CurrentPerm(post_pred_acc, pos, info)
+        any_perm = self.viper.PermGtCmp(post_perm, self.viper.NoPerm(pos, info), pos,
+                                        info)
+
+        # Any subsequently translated assertions (postconditions) are multiplied by this
+        # amount of permission.
+        ctx.perm_factor = post_perm
+
+        object_class = ctx.module.global_module.classes[OBJECT_TYPE]
+        res_var = ctx.actual_function.create_variable('join_result', object_class,
+                                                      self.translator)
+
+        # Resolve list of possible thread target methods.
+        method_options = []
+        for arg in node.args:
+            target = self.get_target(arg, ctx)
+            if not (isinstance(target, PythonMethod) and not target.pure and
+                        not target.predicate):
+                raise InvalidProgramException(node, 'invalid.thread.join')
+            method_options.append(target)
+
+        # Conditionally inhale postconditions of target methods.
+        for method in method_options:
+            stmts.append(self._inhale_possible_thread_post(method, thread, res_var,
+                                                           any_perm, pos, ctx))
+
+        ctx.perm_factor = None
+
+        post_pred = self.viper.PredicateAccessPredicate(post_pred_acc, post_perm, pos,
+                                                        info)
+        exhale_pred = self.viper.Exhale(post_pred, pos, info)
```

Comment:
When you join a thread and you hold a permission to ThreadPost, you get to inhale the postcondition of the thread. If you only have a partial permission to ThreadPost, you'll only get part of the postcondition (e.g., half of all the permissions in it). So when you join and you have a ThreadPost permission, you inhale the postcondition, and you exhale the ThreadPost permission so that you cant join again and get the other half of the permissions, or otherwise get permissions twice.